### PR TITLE
Expose bonding deploy generator as a node sub-command

### PR DESCRIPTION
**Pull request by birchmd**
## Overview
Creates a new sub-command for the node which will generate rholang source code needed for bonding. This simplifies the steps a prospective validator needs to take in order to bond onto the network.

### Which JIRA issue does this PR relate to? If there is not a JIRA issue addressing this work, please create one now and add the link here.
https://rchain.atlassian.net/browse/CORE-1302

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You assigned one person to review this PR

### If you are not a member of the core development team, please confirm:
- [ ] You signed the commit. Merging requires a signature. Please see the [developer wiki](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain) for instructions.
- [ ] Your GitHub account is also an account with [Travis CI](https://travis-ci.org). Unit tests will not run on your PR unless you have an account with Travis. Merge requires passed unit tests.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import coop.rchain.rholang.interpreter.Runtime
+import coop.rchain.shared.PathOps.RichPath
+
+import java.io.PrintWriter
+import java.nio.file.Files
+
+import monix.execution.Scheduler
+
+object BondingUtil {
+  def bondingForwarderAddress(ethAddress: String): String = s"${ethAddress}_bondingForwarder"
```
[the remark this answers is not in the document]
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala]: I am aware of that. I think the `_` was getting included in the variable name for some reason, so I needed to add the braces.

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala [hunk 2]
```diff
@@ -380,5 +381,47 @@ final case class Options(arguments: Seq[String]) extends ScallopConf(arguments)
   }
   addSubcommand(propose)
 
+  val bondingDeployGen = new Subcommand("bdg") {
```
[the remark this answers is not in the document]
Comment (birchmd) [reply node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: I'm open to suggestions here. There is some trade-off between clarity and length though. At first I had it as `bondingDeployGenerate` but then thought that was kind of long for a sub command, so I went full Linux and just took the first letter of each word. There's probably a better middle ground somewhere though.

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala [hunk 3]
```diff
@@ -380,5 +381,47 @@ final case class Options(arguments: Seq[String]) extends ScallopConf(arguments)
...
+    descr(
+      "Creates the rholang source files needed for bonding. These files must be" +
+        "deployed to a node operated by a presently bonded validator. The rho files" +
+        "are created in the working directory where the command is executed. Note: " +
+        "for security reasons it is best to deploy `unlock*.rho` and `forward*.rho` first" +
+        "and `bond*.rho` in a separate block after those."
+    )
+
+    val ethAddr = opt[String](
+      descr = "Ethereum address associated with the \"pre-wallet\" to bond.",
```
[the remark this answers is not in the document]
Comment (birchmd) [reply node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: This is a good point. When I wrote this originally I was imagining people exclusively using the "pre-wallets" that get generated from the wallets file. However, now we have the faucet which creates full basic wallets that use ed25519 keys. Which means for bonding I should write a helper that uses one of those too.

[the remark this answers is not in the document]
Comment (birchmd) [reply node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: I created a ticket for this. I may or may not get to it in this PR. https://rchain.atlassian.net/browse/RHOL-919

File: casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                     .captureResults(runtimeManager.emptyStateHash, unlockSigDataTerm)
+                     .head
+                     .exprs
+                     .head
+                     .getGByteArray
+                     .toByteArray
+                 }
+      unlockSigData = Keccak256.hash(sigBytes)
+      unlockSig     = Secp256k1.sign(unlockSigData, secKey)
+      _             = assert(Secp256k1.verify(unlockSigData, unlockSig, Base16.decode("04" + pubKey)))
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala]: Can you extract out this "04" and add a descriptive label to it?

File: casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      pubKey: String
+  )(implicit scheduler: Scheduler): F[Unit] = {
+    val runtimeDirResource     = makeRuntimeDir[F]
+    val runtimeResource        = makeRuntimeResource[F](runtimeDirResource)
+    val runtimeManagerResource = makeRuntimeManagerResource[F](runtimeResource)
+    runtimeManagerResource.use(
+      runtimeManager =>
+        for {
+          unlockCode <- unlockDeploy[F](ethAddress, pubKey, secKey)(
+                         Sync[F],
```
Comment (KentShikama) [inline casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala]: I thought Sync gets passed in implicitly given how you already are writing `unlockDeploy[F]`

File: node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (KentShikama) [reply node/src/main/scala/coop/rchain/node/configuration/commandline/Options.scala]: I would just call it `generateBondingDeploys`. It is a bit lengthy but I think it is better to be clear than short.

File: casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala (hunk 4 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala]: I don't actually know what the `04` is. I just know that all public keys need to start with `04` in the `Secp256k1` library we use.

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala]: It sounds like we should have at least someone research into why 04 is needed. I know we have this issue in which keys generated in Python for some reason don't work with the node. Maybe it is related?

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/util/BondingUtil.scala]: Yeah probably is related. I bet they would work if you gave them that extra byte

**Pull request merged**